File: oop/coursework/fxsalesman/src/main/java/solver/brutforce/FileManager.java
package solver.brutforce;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class FileManager {
    // Parse input file
    public static Solver parse(String filename) throws IOException {
        File file = new File(filename);
        Scanner sc = new Scanner(file);

        List<List<Integer>> citiesDistances = new ArrayList<List<Integer>>();

        int rowCount = 0;
        int columnCount = 0;
        while(sc.hasNext()){
            String line = sc.nextLine();

            String[] parts = line.split("\\s");
            if (parts.length == 0) {
                continue;
            }

            citiesDistances.add(new ArrayList<Integer>());

            int columnCountTmp = 0;
            for (String part : parts) {
                columnCountTmp += 1;
                try {
                    int distance = Integer.parseInt(part);
                    citiesDistances.get(rowCount).add(distance);
                } catch (NumberFormatException e) {
                    citiesDistances.get(rowCount).add(Solver.passValue);
                    continue;
                }
            }

            if (columnCount == 0) {
                columnCount = columnCountTmp;
            } else if (columnCount != columnCountTmp) {
                throw new IOException("wrong matrix: different column lengths!");
            }

            rowCount++;
        }
        sc.close();

        Solver solver = new Solver(rowCount);
        Matrix.fill(solver, citiesDistances);

        return solver;
    }
}

File: system_programming/cppsalesman/src/FileManager.cpp
#include "FileManager.h"
#include "Solver.h"
#include <fstream>
#include <sstream>
#include <iostream>
#include <regex>

std::vector<std::vector<int>> FileManager::parse(std::string fileName) {
    std::vector<std::vector<int>> citiesDistances;
    int rowsCount = 0;
    int columnsCount = 0;
    std::ifstream infile(fileName);
    std::string line;
    while (std::getline(infile, line)) {
        // Split string
        std::regex rgx("\\s+");
        std::sregex_token_iterator iter(line.begin(), line.end(), rgx, -1);
        std::sregex_token_iterator end;
        if (*iter == line) {
            throw "wrong input data";
        }

        std::vector<int> cityDistances;
        int columnsCountTemp = 0;
        for (; iter != end; ++iter) {
            try {
                int distance = std::stoi(*iter);
                cityDistances.push_back(distance);
            } catch(...) {
                int passValue = Solver::passValue;
                cityDistances.push_back(passValue);
            }
            columnsCountTemp += 1;
        }

        citiesDistances.push_back(cityDistances);

        if (columnsCount == 0) {
            columnsCount = columnsCountTemp;
        } else if (columnsCount != columnsCountTemp) {
            throw  "wrong matrix: different columns lengths!";
        }

        rowsCount += 1;
    }

    return citiesDistances;
}
